### DISCUSSION TITLE
SILE vs ConTeXt vs LyX

### DISCUSSION THREAD

**Discussion by gour** [sile-typesetter/sile#1698] [Q&A] [answered]
Hello,

I did two books in the past using LyX/LaTeX/Xindy..the content had mix of Croatian/English with Sanskrit diacritics - that was the reason to use Xindy for index....few parts, many chapters, appendix, glossary, bibliography, figures (with captions).

Now I'm considering another book project and wonder whether to use markdown/org-mode as source markup and then convert to either LyX and fiddle with LaTeX for a fine-tuning. Another option would be to use ConTeXt which, conceptually, looks nice, but I've never really dived into it, the docs are not very well organized, not certain about its future etc.

SILE looks great, and wonder if I could reproduce the same features as in my old books? What about editor support for its markup? Is it possible to write drafts in some markup and then somehow convert to SILE?

**Comment by Omikhleia** [comment]
>  wonder whether to use markdown (...) as source markup and then convert

SILE has several ways of supporting Markdown, notably natively with [markdown.sile](https://github.com/Omikhleia/markdown.sile) (but other solutions exist too), via Pandoc, etc.

Other than that, it's hard to answer the question, which is very general! Indexes are something where the (current) package might be lacking features.

**Reply by gour** [reply]
> SILE has several ways of supporting Markdown, notably natively with [markdown.sile](https://github.com/Omikhleia/markdown.sile) (but other solutions exist too), via Pandoc, etc.

Thank you. That's good to hear. ;)
>

**Comment by alerque** [comment, marked as answer]
Editor support for the markup only exists [for VIM/NeoVIM](https://github.com/sile-typesetter/vim-sile) as far as I know. It probably wouldn't be hard to work up others, but nobody has done it yet (c.f. #1406 for discussion about vscode support and talk of an LSP).

You can use SILE to typeset directly from Markdown if you want, with or without converting it to SILE's markup first. There is a 3rd party Markdown inputter that parses directly or I have a Pandoc fork I use for conversion first. There are pros and cons of both ways.

Obviously I'd need to see some sample pages to make any more definite statements, but at a guess your "figures with captions" may be the hardest hurdle right now. Not impossible but we don't have off the shelf packages with robust figure placement like LaTeX/ConTeXt have right now. The rest (parts, chapters, appendix, glossary, bibliography, etc.) should be relatively easy to put together.

The mix of languages is not necessarily a problem. There are two approaches though, and I would strongly advise for the somewhat harder approach of tagging your input content. SILE can handle mixed character sets using fallback fonts so you can mix and match UTF-8 input, but there are some language features built into fonts that won't get all the shaping right unless it known the target language at any given point. Hence whether you use SILE native markup like `\language[main=hr]{Hrvatski}`, abbreviate it with a macro like `\hr{Hrvatski}`, or use Markdown like `[Hrvatski]{lang=hr}` (alternatives exist for block/file/etc. level of course, these are inline examples) you'll get better typesetting results if your languages are always explicitly marked up. The same goes for LaTeX/ConTeXt by the way.

If you feel like sharing a couple of page spreads from previous work I or others can probably comment more informatively on how easy or hard it would be to come up with similar results in SILE.

**Reply by gour** [reply]
> Editor support for the markup only exists [for VIM/NeoVIM](https://github.com/sile-typesetter/vim-sile) as far as I know. 

OK. Thanks. In that league, I was also considering Helix editor being modal editor, but with simpler setup than (neo)vim. :-)

> It probably wouldn't be hard to work up others, but nobody has done it yet (c.f. #1406 for discussion about vscode support and talk of an LSP).

Having LSP would, hopefully, solve all editing issues...

> You can use SILE to typeset directly from Markdown if you want, with or without converting it to SILE's markup first. There is a 3rd party Markdown inputter that parses directly or I have a Pandoc fork I use for conversion first. There are pros and cons of both ways.

My gut feeling is that converting to SILE's markup would be preferable, at least for me. What are your pro/cons in regard?

> Obviously I'd need to see some sample pages to make any more definite statements, but at a guess your "figures with captions" may be the hardest hurdle right now. Not impossible but we don't have off the shelf packages with robust figure placement like LaTeX/ConTeXt have right now. 

Actually graphic was quite simple - all the images were on full-colour pictures on their own pages with just captions printed at the bottom, iow. no "robust figure placement".  ;-)

> The rest (parts, chapters, appendix, glossary, bibliography, etc.) should be relatively easy to put together.
Great.

What about index? Can one use `xindy` or `xindex' (not tried that one yet)?

>  The mix of languages is not necessarily a problem. There are two approaches though, and I would strongly advise for the somewhat harder approach of tagging your input content. SILE can handle mixed character sets using fallback fonts so you can mix and match UTF-8 input, but there are some language features built into fonts that won't get all the shaping right unless it known the target language at any given point. 

Good.

> Hence whether you use SILE native markup like `\language[main=hr]{Hrvatski}`, abbreviate it with a macro like `\hr{Hrvatski}`, or use Markdown like `[Hrvatski]{lang=hr}` (alternatives exist for block/file/etc. level of course, these are inline examples) you'll get better typesetting results if your languages are always explicitly marked up. The same goes for LaTeX/ConTeXt by the way.

So, hyphenation is not the problem. Btw, the books I worked on were done long ago in pre-Unicode era taking advantage of (La)TeX accents mechanism and for the Sanskrit words I had to enter many soft-hyphen manually.

> If you feel like sharing a couple of page spreads from previous work I or others can probably comment more informatively on how easy or hard it would be to come up with similar results in SILE.

I'd have to think about something specific and extract from the PDF, but wonder about typing something like:

```
&context

\definefallbackfamily [mainface] [ss] [Noto Sans Devanagari]
   [range={devanagari,devanagariextended},
    features=devanagari-two]

\definefontfamily [mainface] [ss] [Noto Sans]

\setupbodyfont[mainface]

\starttext

ध्यायतो विषयान्पुंसः सङ्गस्तेषूपजायते ।
\blank
सङ्गात्सञ्जायते कामः कामात्क्रोधोऽभिजायते ॥ ६२ ॥

\blank
dhyāyato viṣayān puṁsaḥ
\blank 
saṅgas teṣūpajāyate
\blank
saṅgāt sañjāyate kāmaḥ
\blank
kāmāt krodho ’bhijāyate

\stoptext
```

In the future work I also anticipate using a lot of Astro-related [symbols][https://en.wikipedia.org/wiki/Astronomical_symbols], so I bet that while staying with Unicode symbols that should not be an issue?

I'm on Debian (Sid) and wonder if Ubuntu package is OK for it, or it would be better to be on e.g. Fedora? Writing and publishing is becoming a major focus for me these days, besides note-taking, but for the latter I've decided to use index-cards, so I'm ready to migrate in case I decide for SILE.

**Reply by gour** [reply]
> The same goes for LaTeX/ConTeXt by the way.

What would be your "SILE versus ConTeXt"?

I strongly feel that (for me) it's time to move away from LaTeX (LyX) to something more modern in comparison with fighting against LaTeX's package clashes...

**Reply by alerque** [reply]
Can you post a PDF extract or screen-shoot of a bit of full-featured output?

Also I tried your ConTeXt snippet and don't seem to be getting the Sanskrit bits out properly. It runs but the output seems incomplete.

This would seem to be an equivalent to the snippet you posted using SIL markup:

```sile
\begin{document}
\use[module=packages.font-fallback]
\font[family=Noto Sans]
\font:add-fallback[family=Noto Sans Devanagari]
\set[parameter=document.parskip,value=1bs]
ध्यायतो विषयान्पुंसः सङ्गस्तेषूपजायते ।

सङ्गात्सञ्जायते कामः कामात्क्रोधोऽभिजायते ॥ ६२ ॥

dhyāyato viṣayān puṁsaḥ

saṅgas teṣūpajāyate

saṅgāt sañjāyate kāmaḥ

kāmāt krodho ’bhijāyate
\end{document}
```

![image](https://user-images.githubusercontent.com/173595/215095887-b955e5aa-3502-4993-8de5-4d97ef8c412b.png)

Note this doesn't have any language markup as I would recommend actually doing, but if the specific fonts and language you are using don't rely on script specific features and you don't need hyphenation you can get away with doing things like this. Personally I would setup macros that set the script and font family specifically for each segment, maybe make your document defaults one thing then have macros for each other script you'll use that sets the language, script, and font properties the way you want them.

Another thing to note is ConTeXt is taking me about 1.7 seconds to render your snippet, SILE is taking 0.1 seconds for the equivalent (I'm using LuaJIT, plain Lua will be a bit slower but not THAT much slower).

Anything covered by Unicode should be a non-issue.

I don't know if the Ubuntu packages will work on Debian but they should. If not at least minor tweaks to the package build rules should fix them up for Debian, or you can build from source pretty easily.

**Reply by gour** [reply]
> Can you post a PDF extract or screen-shoot of a bit of full-featured output?

[t.pdf](https://github.com/sile-typesetter/sile/files/10519234/t.pdf)

> This would seem to be an equivalent to the snippet you posted using SIL markup:

It looks nice!

> Note this doesn't have any language markup as I would recommend actually doing, but if the specific fonts and language you are using don't rely on script specific features and you don't need hyphenation you can get away with doing things like this. Personally I would setup macros that set the script and font family specifically for each segment, maybe make your document defaults one thing then have macros for each other script you'll use that sets the language, script, and font properties the way you want them.

OK, that's clear.
 
> Another thing to note is ConTeXt is taking me about 1.7 seconds to render your snippet,


Here is my result:

```
[...]
mkiv lua stats  > runtime: 0.687 seconds, 1 processed pages, 1 shipped pages, 1.455 pages/second
system          | total runtime: 1.381 seconds of 1.428 seconds


________________________________________________________
Executed in    1.43 secs    fish           external
   usr time    1.37 secs    0.00 micros    1.37 secs
   sys time    0.06 secs  460.00 micros    0.06 secs
```

>  SILE is taking 0.1 seconds for the equivalent (I'm using LuaJIT, plain Lua will be a bit slower but not THAT much slower).

That is significant. 

So, using LuaJIT is recommended?

> Anything covered by Unicode should be a non-issue.

Great!

> I don't know if the Ubuntu packages will work on Debian but they should. If not at least minor tweaks to the package build rules should fix them up for Debian, or you can build from source pretty easily.

OK. I'll try to build from the source...

**Reply by alerque** [reply]
> [t.pdf](https://github.com/sile-typesetter/sile/files/10519234/t.pdf)

That's more like I would have expected, I'm not sure why my ConTeXt didn't come up with fonts or something.

> So, using LuaJIT is recommended?

As of v0.14.7 I believe we are at 100% feature parity on LuaJIT vs. PUC Lua. The advantage of LuaJIT is about 2.5× faster document render times, and I've found with OpenResty (a LuaJIT fork) I get about 3× faster than PUC.

The difference is hardly noticeable until you are working on a really long document and fiddling with your document class a lot.

> OK. I'll try to build from the source...

If you're interested in Debian packaging and want to navigate the process of submitting it to Debian (or even just building it privately) I can point you to the debian rules file I keep for the Ubuntu builds and the steps I run to build packages. I don't know how hard it is to navigate the Debian community to be a package maintainer, but since I don't use it locally I've never tried.

**Reply by gour** [reply]
> This would seem to be an equivalent to the snippet you posted using SIL markup:

I took SILE as (another) excuse to migrate to Fedora and here is the result...

[test.pdf](https://github.com/sile-typesetter/sile/files/10529569/test.pdf)

> Another thing to note is ConTeXt is taking me about 1.7 seconds to render your snippet, SILE is taking 0.1 seconds for the equivalent (I'm using LuaJIT, plain Lua will be a bit slower but not THAT much slower).

...along with the timing:

```
$  time sile test.sil 
SILE v0.14.8 (Lua 5.4)
<test.sil> as sil
[1] 

________________________________________________________
Executed in  571.10 millis    fish           external
   usr time  167.62 millis    0.00 micros  167.62 millis
   sys time   63.45 millis  582.00 micros   62.87 millis
```

It looks great!
